### ISSUE TITLE
Tried to convert a mixed precision onnx model to TensorRT FP16 engine

### ISSUE THREAD

**Issue by jinhonglu** [NVIDIA/TensorRT#4019]
## Description

I tried to convert a mixed precision onnx model to mixed precision TensorRT engine.

In my mixed precision onnx model, I have kept some ops (ReduceSum, Pow) to be fp32, some back-to-back Cast Op to be fp32(For example, ReduceSum(fp32)->output(fp32)->Cast(fp32)->Pow(fp32))

In my build_engine.py, I set the following config, set obey precision and set the corresponding layers to be fp32

`        
         config.set_flag(trt.BuilderFlag.FP16)
        config.set_flag(trt.BuilderFlag.OBEY_PRECISION_CONSTRAINTS)

        for i in range(network.num_layers):
            op_name = network.get_layer(i).name.split('/')[-1]
            if 'Pow' == op_name or 'ReduceSum' == op_name or 'Pow_1' == op_name:
                print(network.get_layer(i).name)
                # input('test')
                network.get_layer(i).precision = trt.DataType.FLOAT
                network.get_layer(i).set_output_type(0, trt.DataType.FLOAT)
            if 'Pow_1_output_cast0' == op_name or 'ReduceSum_input_cast1' == op_name or 'Pow_output_cast0' == op_name\
                    or 'Pow_1_input_cast0' == op_name or 'ReduceSum_input_cast0' == op_name or 'Pow_input_cast0' == op_name:
                print(network.get_layer(i).name)
                network.get_layer(i).precision = trt.DataType.FLOAT
`

The result of the tensor engine is quite different from the onnx model.

Any idea that I can solve this?

## Environment

**TensorRT Version**:

**NVIDIA GPU**: A100

**NVIDIA Driver Version**: 12.5

**CUDA Version**:12.5

**CUDNN Version**: 12.5

Operating System: Linux

Python Version (if applicable):

Tensorflow Version (if applicable):

PyTorch Version (if applicable):

Baremetal or Container (if so, version):

## Relevant Files

**Model link**:

## Steps To Reproduce

**Commands or scripts**:

**Have you tried [the latest release](https://developer.nvidia.com/tensorrt)?**:

**Can this model run on other frameworks?** For example run ONNX model with ONNXRuntime (`polygraphy run <model.onnx> --onnxrt`):

**Comment by lix19937** [comment]
What is the diff when you use all fp32 ?

**Comment by jinhonglu** [comment]
> What is the diff when you use all fp32 ?

It is quite strange that the result of the model (onnxfp16->tensortfp32) is also totally different from the onnx fp16.

What is wrong with my build engine code? I have commented the set layer precision at all during building.

**Comment by lix19937** [comment]
You can use  follow to compare the diff between trt and ort.    
```shell  
polygraphy run your_onnx_name.onnx --trt --onnxrt   

```

BTW, if you use trtexec, you can upload the full log with follow cmd  
```shell  
trtexec --verbose --onnx=your_onnx_name.onnx    2>&1    |tee  build.log     

trtexec --verbose --onnx=your_onnx_name.onnx --fp16  2>&1 |tee  build_fp16.log    

```

**Comment by jinhonglu** [comment]
> You can use follow to compare the diff between trt and ort.

`[I] Building engine with configuration:
    Flags                  | [FP16]
    Engine Capability      | EngineCapability.STANDARD
    Memory Pools           | [WORKSPACE: 81157.75 MiB, TACTIC_DRAM: 81157.75 MiB, TACTIC_SHARED_MEMORY: 1024.00 MiB]
    Tactic Sources         | [EDGE_MASK_CONVOLUTIONS, JIT_CONVOLUTIONS]
    Profiling Verbosity    | ProfilingVerbosity.DETAILED
    Preview Features       | [PROFILE_SHARING_0806]
[I] Finished engine building in 96.746 seconds
[I] trt-runner-N0-07/20/24-10:16:56    
    ---- Inference Input(s) ----
    {input_ids [dtype=float16, shape=(1, 2, 1025, 690, 2)]}
[I] trt-runner-N0-07/20/24-10:16:56    
    ---- Inference Output(s) ----
    {output [dtype=float16, shape=(1, 1, 2050, 690, 2)]}
[I] trt-runner-N0-07/20/24-10:16:56     | Completed 1 iteration(s) in 1362 ms | Average inference time: 1362 ms.
[I] onnxrt-runner-N0-07/20/24-10:16:56  | Activating and starting inference
[I] Creating ONNX-Runtime Inference Session with providers: ['CUDAExecutionProvider']
2024-07-20 10:18:42.053428828 [W:onnxruntime:, transformer_memcpy.cc:74 ApplyImpl] 24 Memcpy nodes are added to the graph main_graph for CUDAExecutionProvider. It might have negative impact on performance (including unable to run CUDA graph). Set session_options.log_severity_level=1 to see the detail logs before this message.
2024-07-20 10:18:42.082734265 [W:onnxruntime:, session_state.cc:1166 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.
2024-07-20 10:18:42.082746902 [W:onnxruntime:, session_state.cc:1168 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.
[I] onnxrt-runner-N0-07/20/24-10:16:56 
    ---- Inference Input(s) ----
    {input_ids [dtype=float16, shape=(1, 2, 1025, 690, 2)]}
[I] onnxrt-runner-N0-07/20/24-10:16:56 
    ---- Inference Output(s) ----
    {output [dtype=float16, shape=(1, 1, 2050, 690, 2)]}
[I] onnxrt-runner-N0-07/20/24-10:16:56  | Completed 1 iteration(s) in 253.8 ms | Average inference time: 253.8 ms.
[I] Accuracy Comparison | trt-runner-N0-07/20/24-10:16:56 vs. onnxrt-runner-N0-07/20/24-10:16:56
[I]     Comparing Output: 'output' (dtype=float16, shape=(1, 1, 2050, 690, 2)) with 'output' (dtype=float16, shape=(1, 1, 2050, 690, 2))
[I]         Tolerance: [abs=1e-05, rel=1e-05] | Checking elemwise error
[W]         onnxrt-runner-N0-07/20/24-10:16:56  | Output: output: Some values are 0. Will add a small epsilon quantity to these when computing relative difference. Note that this may cause some relative differences to be extremely high. 
[I]         trt-runner-N0-07/20/24-10:16:56: output | Stats: mean=0.11066, std-dev=0.11466, var=0.013147, median=0.14856, min=-0.05246 at (0, 0, 2047, 264, 1), max=1.1465 at (0, 0, 2048, 678, 0), avg-magnitude=0.11456
[I]             ---- Histogram ----
                Bin Range          |  Num Elems | Visualization
                (-0.513 , -0.347 ) |          0 | 
                (-0.347 , -0.181 ) |          0 | 
                (-0.181 , -0.0149) |      94163 | ##
                (-0.0149, 0.151  ) |    1322296 | #####################################
                (0.151  , 0.317  ) |    1393786 | ########################################
                (0.317  , 0.483  ) |      16685 | 
                (0.483  , 0.649  ) |       1380 | 
                (0.649  , 0.814  ) |          0 | 
                (0.814  , 0.98   ) |          0 | 
                (0.98   , 1.15   ) |        690 | 
[I]         onnxrt-runner-N0-07/20/24-10:16:56: output | Stats: mean=0.00063173, std-dev=0.013393, var=0.00017937, median=0, min=-0.5127 at (0, 0, 2048, 531, 1), max=0.60645 at (0, 0, 0, 19, 0), avg-magnitude=0.0012219
[I]             ---- Histogram ----
                Bin Range          |  Num Elems | Visualization
                (-0.513 , -0.347 ) |        690 | 
                (-0.347 , -0.181 ) |          2 | 
                (-0.181 , -0.0149) |      10507 | 
                (-0.0149, 0.151  ) |    2814857 | ########################################
                (0.151  , 0.317  ) |       2242 | 
                (0.317  , 0.483  ) |        692 | 
                (0.483  , 0.649  ) |         10 | 
                (0.649  , 0.814  ) |          0 | 
                (0.814  , 0.98   ) |          0 | 
                (0.98   , 1.15   ) |          0 | 
[I]         Error Metrics: output
[I]             Minimum Required Tolerance: elemwise error | [abs=1.1464] OR [rel=1.4492e+15] (requirements may be lower if both abs/rel tolerances are set)
[I]             Absolute Difference | Stats: mean=0.11426, std-dev=0.11022, var=0.012149, median=0.088135, min=0 at (0, 0, 13, 309, 1), max=1.1464 at (0, 0, 2048, 678, 0), avg-magnitude=0.11426
[I]                 ---- Histogram ----
                    Bin Range      |  Num Elems | Visualization
                    (0    , 0.115) |    1416722 | ########################################
                    (0.115, 0.229) |     915471 | #########################
                    (0.229, 0.344) |     490002 | #############
                    (0.344, 0.459) |       4044 | 
                    (0.459, 0.573) |       2071 | 
                    (0.573, 0.688) |          0 | 
                    (0.688, 0.802) |          0 | 
                    (0.802, 0.917) |          0 | 
                    (0.917, 1.03 ) |          0 | 
                    (1.03 , 1.15 ) |        690 | 
[I]             Relative Difference | Stats: mean=8.9299e+13, std-dev=2.6135e+14, var=6.8306e+28, median=1.2531e+12, min=0 at (0, 0, 13, 309, 1), max=1.4492e+15 at (0, 0, 121, 183, 0), avg-magnitude=8.9299e+13
[I]                 ---- Histogram ----
                    Bin Range            |  Num Elems | Visualization
                    (0       , 1.45e+14) |    2608328 | ########################################
                    (1.45e+14, 2.9e+14 ) |       4871 | 
                    (2.9e+14 , 4.35e+14) |          0 | 
                    (4.35e+14, 5.8e+14 ) |          0 | 
                    (5.8e+14 , 7.25e+14) |       5599 | 
                    (7.25e+14, 8.69e+14) |      33519 | 
                    (8.69e+14, 1.01e+15) |      90300 | #
                    (1.01e+15, 1.16e+15) |      68294 | #
                    (1.16e+15, 1.3e+15 ) |      15648 | 
                    (1.3e+15 , 1.45e+15) |       2441 | `

**Comment by LeoZDong** [comment]
Could you provide your ONNX model file?

> It is quite strange that the result of the model (onnxfp16->tensortfp32) is also totally different from the onnx fp16.
Are you saying if you convert your mixed precision model into FP16 in ONNX, the output looks very different from your original mixed precision model?

**Comment by jinhonglu** [comment]
I fixed the problem with polygraphy debug precision

Thank you.

**Comment by JohnHerry** [comment]
Is there any detailed solution infomation? @jinhonglu
